# After you've done your QC
# Input files:
# QCed plink bim file
# also, calculate MAF using --freq using plink
# (not sure how helpful this'll be for us given the small sample size - likely to figure out allele flips)

# https://www.well.ox.ac.uk/~wrayner/tools/
# The tool is a perl script, and you should be able to run the one under my home dir

# Usage with the TOPMed reference panel
# NOTE: It is recommended to use v4.3.0 for the TOPMed panel,
# previous versions will work with the TOPMed panel but will require around 300GB RAM.
# v4.3.0 reduces the memory usage substantially and this version can be used with all other reference panels as well,
# with a corresponding reduction in memory usage.

# The TOPMed reference panel is not available for direct download from this site,
# it needs to be created from the VCF of dbSNP submitted sites (currently ALL.TOPMed_freeze5_hg38_dbSNP.vcf.gz).
# This can be downloaded from the Bravo website: https://bravo.sph.umich.edu/freeze5/hg38/
# Build 37 can also be downloaded from the Bravo website: 

# Once downloaded the VCF can be converted to an HRC formatted reference legend using the code: CreateTOPMed.zip
# Usage: ./CreateTOPMed.pl -i ALL.TOPMed_freeze5_hg38_dbSNP.vcf.gz

# By default this will create a file filtered for variants flagged as PASS only, if you wish to use all variants the -a flag overrides this.
# To override the default output file naming use -o filename.

# To run the checks the usage is the same as for the HRC panel, namely using the flags -h -r PASS.Variants.TOPMed_freeze5_hg38_dbSNP.tab.gz
# Usage: perl HRC-1000G-check-bim.pl -b <bim file> -f <Frequency file> -r <Reference panel> -h

# HRC imputation panel:
# Requires the unzipped tab delimited HRC reference (currently v1.1 HRC.r1-1.GRCh37.wgs.mac5.sites.tab) 
# from the Haplotype Reference Consortium Website: http://www.haplotype-reference-consortium.org/site

cd /well/lindgren/UKBIOBANK/dpalmer/PRS_cell_data

# Choose reference panel:
RefPanel="data/HRC_variants/HRC.r1-1.GRCh37.wgs.mac5.sites.tab.gz"
export PATH="/well/lindgren/dpalmer/:$PATH"

plink --bfile data/celldataB37/celldataB37_dp_test --freq --out data/celldataB37/celldataB37_dp_test

# Run the check tool
perl HRC-1000G-check-bim.pl \
    -b data/celldataB37/celldataB37_dp_test.bim \
    -f data/celldataB37/celldataB37_dp_test.frq \
    -t 0.2 -r ${RefPanel} -h -o data/celldataB37/HRC

# The suggested plink commands generated by the tool are in this "Run-plink.sh" file
cd data/celldataB37/HRC
head -n 5 Run-plink.sh
mv Run-plink.sh Run-plink-B37.sh
bash Run-plink-B37.sh
rm /well/lindgren/UKBIOBANK/dpalmer/PRS_cell_data/data/celldataB37/TEMP*

# The updated plink files
wc -l celldataB37_dp_test-updated.bim

module purge
module load BCFtools/1.12-GCC-9.3.0
for chr in {1..23}
do
    bgzip celldataB37_dp_test-updated-chr${chr}.vcf
done

# These files are now ready to impute using the Michigan imputation server.
to_impute="/well/lindgren/UKBIOBANK/dpalmer/PRS_cell_data/data/celldataB37/HRC"
# Before running, ensure you have a valid API token: see https://github.com/lukfor/imputationbot.
imputationbot impute \
    --files ${to_impute}/*vcf.gz \
    --refpanel hrc-r1.1 \
    --population mixed \
    --r2Filter 0.3 \
    --autoDownload \
    --password ${mypassword}

mkdir /well/lindgren/UKBIOBANK/dpalmer/PRS_cell_data/data/celldataB37/HRC/imputed
mv chr*zip /well/lindgren/UKBIOBANK/dpalmer/PRS_cell_data/data/celldataB37/HRC/imputed/

# We will only use the HRC imputed data because UK Biobank is imputed to HRC, but include
# instructions to prepare data for imputation to TOPMED for completeness.

# TOPMED imputation panel:

cd /well/lindgren/UKBIOBANK/dpalmer/PRS_cell_data
mkdir data/for_imputation/

# Build 37
perl CreateTOPMed.pl \
    -i data/TOPMed_variants/bravo-dbsnp-all.GRCh37.vcf.gz \
    -o data/for_imputation/PASS.topmed.GRCh37.tab.gz

# Choose reference panel:
RefPanel=data/for_imputation/PASS.topmed.GRCh37.tab.gz
export PATH="/well/lindgren/dpalmer/:$PATH"

plink --bfile data/celldataB37_dp_test --freq --out data/celldataB37_dp_test

# Run the check tool
perl HRC-1000G-check-bim.pl \
    -b data/celldataB37_dp_test.bim \
    -f data/celldataB37_dp_test.frq \
    -t 0.2 -r ${RefPanel} -h

# The suggested plink commands generated by the tool are in this "Run-plink.sh" file
cd data
head -n 5 Run-plink.sh
mv Run-plink.sh Run-plink-B37.sh
bash Run-plink-B37.sh

# The updated plink files
wc -l celldataB37_dp_test-updated.bim

module purge
module load BCFtools/1.12-GCC-9.3.0
for chr in {1..23}
do
    bgzip celldataB37_dp_test-updated-chr${chr}.vcf
done

mkdir celldataB37
mv *B37* celldataB37/

# Build 38
cd ..
perl CreateTOPMed.pl \
    -i data/TOPMed_variants/bravo-dbsnp-all.GRCh38.vcf.gz \
    -o data/for_imputation/PASS.topmed.GRCh38.tab.gz

# Choose reference panel:
RefPanel=data/for_imputation/PASS.topmed.GRCh38.tab.gz
export PATH="/well/lindgren/dpalmer/:$PATH"

plink --bfile data/celldataB38_test --freq --out data/celldataB38_test

# Run the check tool
perl HRC-1000G-check-bim.pl \
    -b data/celldataB38_test.bim \
    -f data/celldataB38_test.frq \
    -t 0.2 -r ${RefPanel} -h

# The suggested plink commands generated by the tool are in this "Run-plink.sh" file
cd data
head -n 5 Run-plink.sh
mv Run-plink.sh Run-plink-B38.sh
bash Run-plink-B38.sh

# The updated plink files
wc -l celldataB38_test-updated.bim

module purge
module load BCFtools/1.12-GCC-9.3.0
for chr in {1..23}
do
    awk '{if($0 !~ /^#/) print "chr"$0; else print $0}' celldataB38_test-updated-chr${chr}.vcf \
    > tmp && mv tmp celldataB38_test-updated-chr${chr}.vcf
    bgzip celldataB38_test-updated-chr${chr}.vcf
done

mkdir celldataB38
mv *B38* celldataB38/

mkdir celldataB36
mv *B36* celldataB36/

