# After you've done your QC
# Input files:
# QCed plink bim file
# also, calculate MAF using --freq using plink
# (not sure how helpful this'll be for us given the small sample size - likely to figure out allele flips)

# https://www.well.ox.ac.uk/~wrayner/tools/
# The tool is a perl script, and you should be able to run the one under my home dir

# Usage with the TOPMed reference panel
# NOTE: It is recommended to use v4.3.0 for the TOPMed panel,
# previous versions will work with the TOPMed panel but will require around 300GB RAM.
# v4.3.0 reduces the memory usage substantially and this version can be used with all other reference panels as well,
# with a corresponding reduction in memory usage.

# The TOPMed reference panel is not available for direct download from this site,
# it needs to be created from the VCF of dbSNP submitted sites (currently ALL.TOPMed_freeze5_hg38_dbSNP.vcf.gz).
# This can be downloaded from the Bravo website: https://bravo.sph.umich.edu/freeze5/hg38/
# Build 37 can also be downloaded from the Bravo website: 

# Once downloaded the VCF can be converted to an HRC formatted reference legend using the code: CreateTOPMed.zip
# Usage: ./CreateTOPMed.pl -i ALL.TOPMed_freeze5_hg38_dbSNP.vcf.gz

# By default this will create a file filtered for variants flagged as PASS only, if you wish to use all variants the -a flag overrides this.
# To override the default output file naming use -o filename.

# To run the checks the usage is the same as for the HRC panel, namely using the flags -h -r PASS.Variants.TOPMed_freeze5_hg38_dbSNP.tab.gz
# Usage: perl HRC-1000G-check-bim.pl -b <bim file> -f <Frequency file> -r <Reference panel> -h

cd /well/lindgren/UKBIOBANK/dpalmer/PRS_cell_data
mkdir data/for_imputation/

# Build 37
perl CreateTOPMed.pl -i data/TOPMed_variants/bravo-dbsnp-all.GRCh37.vcf.gz -o data/for_imputation/PASS.topmed.GRCh37.tab.gz

# Choose reference panel:
RefPanel=data/for_imputation/PASS.topmed.GRCh37.tab.gz
export PATH="/well/lindgren/dpalmer/:$PATH"

# Merge them all together

for pop in EUR AFR AMR EAS SAS
do
    plink --bfile data/UKB_subset_${pop}_test --freq --out data/UKB_subset_${pop}_test

    # Run the check tool
    # -t default is 0.2 but I set it to 1 so that no variants are removed because have have so few samples
    # Submit a job (RAM ~60G)
    if [ $pop == EUR ]; then
        perl HRC-1000G-check-bim.pl -b data/UKB_subset_${pop}_test.bim -f data/UKB_subset_${pop}_test.frq -t 0.2 -r ${RefPanel} -h
    else
        perl HRC-1000G-check-bim.pl -b data/UKB_subset_${pop}_test.bim -f data/UKB_subset_${pop}_test.frq -t 0.5 -r ${RefPanel} -h
    fi
    # The suggested plink commands generated by the tool are in this "Run-plink.sh" file
    cd data
    head -n 5 Run-plink.sh
    mv Run-plink.sh Run-plink-B37-ukb-${pop}.sh
    bash Run-plink-B37-ukb-${pop}.sh

    # The updated plink files
    wc -l UKB_subset_${pop}_test-updated.bim
    cd ..

done

module purge
module load BCFtools/1.12-GCC-9.3.0
cd data

for pop in EUR AFR AMR EAS SAS
do
    mkdir ${pop}
    for chr in {1..23}
    do
        bgzip UKB_subset_${pop}_test-updated-chr${chr}.vcf
        mv UKB_subset_${pop}_test-updated-chr${chr}.vcf.gz ${pop}/
    done
done

for pop in EUR AFR AMR EAS SAS
do
    mv *${pop}* ${pop}/
done
